cdap-docs/developers-manual/source/building-blocks/transaction-system.rst

```diff
@@ -158,3 +160,189 @@ It's important to note that the MapReduce framework will reattempt a task (Mappe
+these must be together in the same transaction and committed atomically: otherwise,
+exactly-once processing cannot be ensured.
+
+For other types of programs, transactions can also be useful. For example, the handler
+methods of services are executed transactionally to make sure they operate on consistent
+data. The lifecycle methods (``initialize()`` and ``destroy()``) of all programs are also executed
+within an implicit transaction.
+
+However, there are use cases where that transaction is not desired:
+
+- The default transaction timeout (as :ref:`configured by
+  <appendix-cdap-default-datasets>` ``data.tx.timeout`` in ``cdap-site.xml``) may be too
+  short for the operations performed by a method. For example, the ``destroy()`` method of a
+  MapReduce program may have to clean up temporary data, or make a web service call to
+  notify some other party of the job completion.
+
+- A method does not perform any transactional operations. For example, FileSet datasets do
+  not require a transaction |---| a method using only FileSets therefore does not require an 
+  implicit transaction.
+
+- A method performs many operations and wishes to execute them in several short
+  transactions rather than a single long transaction. A good example of such a method is the
+  ``run()`` method of a worker, which runs perpetually and cannot be executed inside a single
+  transaction. Instead, it needs to start an explicit transaction whenever it performs
+  operations on transactional datasets. 
+
+To facilitate these use cases, CDAP offers programs control over the execution of
+transactions:
+
+- Annotate a method with an ``@TransactionPolicy`` to turn off the implicit transaction
+  started by CDAP.
+
+- Use the program context’s ``execute()`` method to run a block of code inside an explicit
+  transaction.
+
+- Control the timeout of transactions by setting a system-wide configuration
+  (``data.tx.timeout``); by setting a preference for an individual namespace, application, or
+  program; or by passing a timeout for the transaction to the ``execute()`` method.
+
+Implicit versus Explicit Transactions
+-------------------------------------
+By default, CDAP will start an implicit transaction for these methods:
+
+- All flowlet process methods
+- All service handler methods
+- The ``ProgramLifecycle`` methods (``initialize()`` and ``destroy()``) for all types of
+  programs and sub-programs (flowlets, service handlers, and workflow actions), with the
+  exception of worker programs.
+
+For flowlet process methods, this cannot be disabled, because that would impact the
+semantics of flow execution. For MapReduce programs, the lifecycle methods of MapReduce
+tasks (mappers and reducers) and MapReduce helpers (such as partitioners and comparators)
+are always run inside a transaction: the long-running transaction that encapsulates an
+entire MapReduce job (see :ref:`above <transaction-system-transactions-mapreduce>`). 
+
```

Comment:
Is this a good place to mention: Please see [link to spark section] for how to use transactions in Spark?